hydromt_delft3dfm/dflowfm.py

```diff
@@ -1487,6 +1458,385 @@ def setup_1dboundary(
             da_out, name=f"boundary1d_{da_out.name}_{branch_type}"
         )  # FIXME: this format cannot be read back due to lack of branch type info from model files
 
+    def _read_forcing_geodataset(
+        self, forcing_geodataset_fn: Union[str, Path], forcing_name: str = "discharge"
+    ):
+        """Read forcing geodataset."""
+
+        refdate, tstart, tstop = self.get_model_time()  # time slice
+
+        if (
+            forcing_geodataset_fn is not None
+            and self.data_catalog[forcing_geodataset_fn].data_type == "GeoDataset"
+        ):
+            da = self.data_catalog.get_geodataset(
+                forcing_geodataset_fn,
+                geom=self.region.buffer(
+                    1000
+                ),  # only select data within region of interest (large region for forcing)
+                variables=[forcing_name],
+                time_tuple=(tstart, tstop),
+            ).rename(forcing_name)
+            # error if time mismatch
+            if np.logical_and(
+                pd.to_datetime(da.time.values[0]) == pd.to_datetime(tstart),
+                pd.to_datetime(da.time.values[-1]) == pd.to_datetime(tstop),
+            ):
+                pass
+            else:
+                self.logger.error(
+                    "forcing has different start and end time. Please check the forcing file. support yyyy-mm-dd HH:MM:SS. "
+                )
+            # reproject if needed and convert to location
+            if da.vector.crs != self.crs:
+                da = da.vector.to_crs(self.crs)
+            # get geom
+            gdf = da.vector.to_gdf(reducer=np.mean)
+        elif (
+            forcing_geodataset_fn is not None
+            and self.data_catalog[forcing_geodataset_fn].data_type == "GeoDataFrame"
+        ):
+            gdf = self.data_catalog.get_geodataframe(
+                forcing_geodataset_fn,
+                geom=self.region.buffer(self._network_snap_offset),
+            )
+            da = None
+        else:
+            gdf = None
+            da = None
+        return gdf, da
+
+    def _snap_geom_to_branches_and_drop_nonsnapped(
```

Comment:
Can this function be in a workflow instead ? (self is not used)

Reply:
Yes!